src/main/java/com/epam/ta/reportportal/core/launch/impl/StartLaunchHandlerImpl.java

```diff
@@ -68,22 +76,26 @@ public StartLaunchRS startLaunch(ReportPortalUser user, ReportPortalUser.Project
 		Long userId = userRepository.findIdByLoginForUpdate(user.getUsername())
 				.orElseThrow(() -> new ReportPortalException(ErrorType.USER_NOT_FOUND, user.getUsername()));
 
-		if (request.isRerun()) {
-			return rerunHandler.handleLaunch(request, projectDetails.getProjectId(), user);
-		}
+		final Launch savedLaunch = Optional.of(request.isRerun())
+				.filter(Boolean::booleanValue)
+				.map(rerun -> rerunHandler.handleLaunch(request, projectDetails.getProjectId(), user))
+				.orElseGet(() -> {
+					Launch launch = new LaunchBuilder().addStartRQ(request)
+							.addAttributes(request.getAttributes())
+							.addProject(projectDetails.getProjectId())
+							.addUserId(userId)
+							.get();
+					launchRepository.save(launch);
+					launchRepository.refresh(launch);
+					return launch;
+				});
 
-		Launch launch = new LaunchBuilder().addStartRQ(request)
-				.addAttributes(request.getAttributes())
-				.addProject(projectDetails.getProjectId())
-				.addUserId(userId)
-				.get();
-		launchRepository.save(launch);
-		launchRepository.refresh(launch);
-		messageBus.publishActivity(new LaunchStartedEvent(TO_ACTIVITY_RESOURCE.apply(launch), user.getUserId(), user.getUsername()));
+		eventPublisher.publishEvent(new LaunchEvent(savedLaunch.getId(), START_KEY));
```

Comment:
Can it be a separate class for this type of event?

Reply:
replaced